File: src/main/java/ux/ColorConstant.java
package ux;

import processing.core.PApplet;

public class ColorConstant {
    private int currentColor;
    private int previousColor;
    private long transitionStartTime;
    private static final int TRANSITION_DURATION = UXThemeManager.getInstance().getThemeTransitionDuration();  // 2 seconds
    private boolean transitionInProgress = false;
    private PApplet p;
    private boolean isInitialized = false;

    public ColorConstant() {}

    public void setColor(int newColor, PApplet p) {
        if (isInitialized) {
            this.previousColor = this.currentColor;
            this.transitionInProgress = true;
            this.transitionStartTime = System.currentTimeMillis();
            this.p = p;
        } else {
            this.previousColor = newColor;
            this.isInitialized = true;
        }
        this.currentColor = newColor;
    }

    public int getColor() {
        if (p == null || !transitionInProgress) {
            return currentColor;
        }

        float t = (float) (System.currentTimeMillis()- transitionStartTime) / TRANSITION_DURATION;
        if (t < 1.0) {
            return p.lerpColor(previousColor, currentColor, t);
        } else {
            transitionInProgress = false;
            return currentColor;
        }
    }
}

File: src/main/kotlin/ux/ColorConstant.kt
package ux

import processing.core.PApplet

class ColorConstant(private val p: PApplet) {
    private var currentColor = 0
    private var previousColor = 0
    private var transitionStartTime: Long = 0
    private var transitionInProgress = false
    private var isInitialized = false

    fun setColor(newColor: Int) {
        if (isInitialized) {
            previousColor = currentColor
            transitionInProgress = true
            transitionStartTime = System.currentTimeMillis()
        } else {
            previousColor = newColor
            this.isInitialized = true
        }
        currentColor = newColor
    }

    val color: Int
        get() {
            if (!transitionInProgress) {
                return currentColor
            }
            val t = (System.currentTimeMillis() - transitionStartTime).toFloat() / Theme.themeTransitionDuration
            return if (t < 1.0) {
                p.lerpColor(previousColor, currentColor, t)
            } else {
                transitionInProgress = false
                currentColor
            }
        }
}